crates/interfaces/src/test_utils/full_block.rs

```diff
@@ -87,80 +112,120 @@ impl TestFullBlockClient {
+        // Retrieves headers based on the provided limit and request direction.
+        let resp = (0..request.limit)
+            .filter_map(|_| {
+                headers.iter().find_map(|(hash, header)| {
+                    // Checks if the header matches the specified block or number.
+                    if BlockNumHash::new(
+                        header.number,
+                        reth_primitives::alloy_primitives::FixedBytes(**hash),
+                    )
+                    .matches_block_or_num(&block)
+                    {
+                        match request.direction {
+                            HeadersDirection::Falling => block = header.parent_hash.into(),
+                            HeadersDirection::Rising => block = (header.number + 1).into(),
+                        }
+                        Some(header.clone())
+                    } else {
+                        None
                     }
-                }
-                resp.push(header.clone());
-            } else {
-                break
-            }
-        }
+                })
+            })
+            .collect::<Vec<_>>();
+
+        // Returns a future containing the retrieved headers with a random peer ID.
         futures::future::ready(Ok(WithPeerId::new(PeerId::random(), resp)))
     }
 }
 
+/// Implements the `BodiesClient` trait for the `TestFullBlockClient` struct.
 impl BodiesClient for TestFullBlockClient {
+    /// Defines the output type of the function.
     type Output = futures::future::Ready<PeerRequestResult<Vec<BlockBody>>>;
 
+    /// Retrieves block bodies corresponding to provided hashes with a given priority.
+    ///
+    /// # Arguments
+    ///
+    /// * `hashes` - A vector of block hashes to retrieve bodies for.
+    /// * `_priority` - Priority level for block body retrieval (unused in this implementation).
+    ///
+    /// # Returns
+    ///
+    /// A future containing the result of the block body retrieval operation.
     fn get_block_bodies_with_priority(
         &self,
         hashes: Vec<B256>,
         _priority: Priority,
     ) -> Self::Output {
+        // Acquire a lock on the bodies.
         let bodies = self.bodies.lock();
-        let mut all_bodies = Vec::new();
-        for hash in hashes {
-            if let Some(body) = bodies.get(&hash) {
-                all_bodies.push(body.clone());
-            }
-
-            if all_bodies.len() == self.soft_limit {
-                break
-            }
-        }
-        futures::future::ready(Ok(WithPeerId::new(PeerId::random(), all_bodies)))
+
+        // Create a future that immediately returns the result of the block body retrieval
+        // operation.
+        futures::future::ready(Ok(WithPeerId::new(
+            PeerId::random(),
+            hashes
+                .iter()
+                .filter_map(|hash| bodies.get(hash).cloned())
+                .take(self.soft_limit)
+                .collect::<Vec<_>>(),
```

Comment:
is the type hint required here, maybe we don't even need that

Reply:
nice, fixed